File: ga/gpt3_5_result_2024_6/result/1968G2_gpt/29_1968G2.c
use std::cmp;

const N: usize = 200000;
const N_: usize = 1 << 18; // N_ = 2^ceil(log2(N))

fn max(a: i32, b: i32) -> i32 {
    if a > b {
        a
    } else {
        b
    }
}

fn build(st: &mut [i32], aa: &[i32]) {
    let mut n_ = 1;
    while n_ < aa.len() {
        n_ <<= 1;
    }
    st.iter_mut().for_each(|x| *x = 0);
    aa.iter().enumerate().for_each(|(i, &val)| st[n_ + i] = val);
    for i in (1..n_).rev() {
        st[i] = max(st[i << 1], st[(i << 1) + 1]);
    }
}

fn query(st: &[i32], n_: usize, l: usize, a: i32) -> usize {
    let mut r = n_ - 1;
    let mut l_val = l;
    
    let mut l = l + n_;
    let mut r = r + n_;
    while l <= r {
        if l & 1 == 1 {
            if st[l] >= a {
                while l < n_ {
                    l = if st[l << 1] >= a { l << 1 } else { (l << 1) + 1 };
                }
                return l - n_;
            }
            l += 1;
        }
        l >>= 1;
        r >>= 1;
    }
    n_
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: i32 = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: i32 = iter.next().unwrap().parse().unwrap();
        let kl: i32 = iter.next().unwrap().parse().unwrap();
        let kr: i32 = iter.next().unwrap().parse().unwrap();
        let cc: String = iter.next().unwrap().to_string();

        let mut zz = vec![0; n as usize];
        let cc_chars: Vec<char> = cc.chars().collect();

        let mut l = 0;
        let mut r = 0;

        for i in 1..n {
            if i + zz[i as usize - l as usize] < r {
                zz[i as usize] = zz[i as usize - l as usize];
            } else {
                l = i;
                r = cmp::max(r, i);
                while r < n as usize && cc_chars[r] == cc_chars[r - l] {
                    r += 1;
                }
                zz[i as usize] = (r - l) as i32;
            }
        }

        let mut st = vec![0; N_ * 2];
        build(&mut st, &zz);

        let mut ans = vec![0; (n as usize + 1)];
        for p in 1..=n as usize {
            let mut k = 0;
            let mut i = 0;
            while i < n as usize {
                i = query(&st, N_, i + p, p as i32);
                if i < n as usize {
                    k += 1;
                }
            }
            ans[k] = cmp::max(ans[k], p);
        }

        for k in (1..=n as usize - 1).rev() {
            ans[k] = cmp::max(ans[k], ans[k + 1]);
        }

        for k in kl as usize..=kr as usize {
            print!("{} ", ans[k]);
        }
        println!();
    }
}

File: ga/gpt3_5_result_2024_6/result/1968G2_cpp_gpt/29_1968G2.rs
use std::cmp;
use std::collections::HashMap;

fn find(k: usize, p: &mut Vec<usize>) -> usize {
    if k > p.len() || p[k] == k {
        return k;
    }
    p[k] = find(p[k], p);
    p[k]
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: i32 = input.trim().parse().unwrap();
    
    for _ in 0..t {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let x: usize = iter.next().unwrap().parse().unwrap();
        let y: usize = iter.next().unwrap().parse().unwrap();
        
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let a: Vec<char> = input.trim().chars().collect();
        
        let mut z: Vec<usize> = vec![0; n+1];
        let mut p: Vec<usize> = (0..=n+1).collect();
        let mut t: Vec<usize> = (0..=n+1).collect();
        let mut ans: Vec<usize> = vec![0; n+1];
        
        let mut j = 0;
        z[1] = n;
        for (i, ch) in a.iter().enumerate().skip(1) {
            if j + z[j] > i {
                z[i] = cmp::min(z[i-j+1], j + z[j] - i);
            }
            while a[i+z[i]] == a[1+z[i]] {
                z[i] += 1;
            }
            if i + z[i] > j + z[j] {
                j = i;
            }
        }
        
        t.sort_by(|&i, &j| z[i].cmp(&z[j]));
        
        let mut k = 1;
        for i in 1..=n {
            let mut c = 0;
            while k <= n && z[t[k]] < i {
                p[t[k]] = t[k] + 1;
                k += 1;
            }
            let mut j = 1;
            while j <= n {
                let root = find(j + i, &mut p);
                c += 1;
                j = root;
            }
            ans[c] = i;
        }
        
        for i in (1..=n).rev() {
            ans[i] = cmp::max(ans[i], ans[i+1]);
        }
        
        for i in x..=y {
            print!("{} ", ans[i]);
        }
        println!("");
    }
}